### PR TITLE
PCHR-1304: Implement edit mode for leave request

### PR DESCRIPTION
**Pull request by dhavalsharma**
The edit mode for leave request is implemented. The gif is attached here

edit-leave-request

## Implementation
A local variable `initialLeaveRequest` is added to track changes to leave request object. It is used to check if object is changed and thereafter allow user to submit the changes. 

The `initDates` function had to return promise in order to set the `initialLeaveRequest` correctly.

Additional event `LeaveRequest::edit` is added to handle edit mode.

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 1]
```diff
@@ -155,6 +156,10 @@ define([
       $rootScope.$on('LeaveRequest::new', function () {
         vm.refresh();
       });
+
+      $rootScope.$on('LeaveRequest::edit', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Can you group all the `$rootScope.$on` calls in a separate function? Would make the code better organized

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 2]
```diff
@@ -201,6 +202,11 @@ define([
       vm.canSubmit = function () {
         var canSubmit = canCalculateChange();
 
+        //check if user has changed any attribute
+        if (vm.mode == 'edit') {
+          canSubmit = canSubmit &&  !_.isEqual(initialLeaveRequest, vm.leaveRequest.attributes());
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: Is this a requirement in terms of UX? Why not allowing the user to re-save the leave request as is, if he wants to?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 3]
```diff
@@ -285,7 +291,14 @@ define([
             initAbsenceType();
             initStatus();
             initContact();
-            initDates();
+
+            //the promise will set the day types so that initial value is set properly
+            return initDates();
+          })
+          .then( function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]:
```js
.then(function () {
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 4]
```diff
@@ -285,7 +291,14 @@ define([
...
+            if(vm.mode == 'edit') {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]:
```js
if (vm.mode === 'edit') {
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 5]
```diff
@@ -285,7 +291,14 @@ define([
...
+              initialLeaveRequest = _.cloneDeep(vm.leaveRequest.attributes());
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: You don't need to clone the result of `attributes()`, given that those attributes are stored in a brand new object
```js
attributes: function () {
    return _.transform(this, function (result, __, key) {
        !_.isFunction(this[key]) && (result[key] = this[key]);
    }, Object.create(null), this);
},
```

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: Also I think the name of this variable is incorrect, as it's only containing the attributes of the leave request, not the leave request per se

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 6]
```diff
@@ -735,21 +748,35 @@ define([
...
       /**
        * Initialize from and to dates, day types and contact
+       *
+       * @return {Promise}
        */
       function initDates() {
+        var deferred = $q.defer();
+
         if (canEdit()) {
           var attributes = _.cloneDeep(vm.leaveRequest.attributes());
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: You already have this in `initialLeaveRequest`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 7]
```diff
@@ -735,21 +748,35 @@ define([
...
+
           vm.onDateChange(vm.uiOptions.fromDate, 'from')
             .then(function () {
               //to_date and type has been reset in above call so reinitialize from clone
               vm.leaveRequest.to_date = attributes.to_date;
               vm.leaveRequest.to_date_type = attributes.to_date_type;
               vm.uiOptions.toDate = convertDateFormatFromServer(vm.leaveRequest.to_date);
-              vm.onDateChange(vm.uiOptions.toDate, 'to');
+              vm.onDateChange(vm.uiOptions.toDate, 'to')
+                .then( function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]:
```js
.then(function () {
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 8]
```diff
@@ -829,7 +856,13 @@ define([
       function updateRequest() {
         vm.leaveRequest.update()
           .then(function () {
-            postSubmit('LeaveRequest::updatedByManager');
+            if (vm.role == 'manager') {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: `===`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 9]
```diff
@@ -735,21 +748,35 @@ define([
 
       /**
        * Initialize from and to dates, day types and contact
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: why *and contact*?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 10]
```diff
@@ -735,21 +748,35 @@ define([
...
        */
       function initDates() {
+        var deferred = $q.defer();
+
         if (canEdit()) {
           var attributes = _.cloneDeep(vm.leaveRequest.attributes());
 
           vm.uiOptions.fromDate = convertDateFormatFromServer(vm.leaveRequest.from_date);
+
           vm.onDateChange(vm.uiOptions.fromDate, 'from')
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: You still have this `onSomething` methods. Please change the name

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 11]
```diff
@@ -735,21 +748,35 @@ define([
...
             .then(function () {
               //to_date and type has been reset in above call so reinitialize from clone
               vm.leaveRequest.to_date = attributes.to_date;
               vm.leaveRequest.to_date_type = attributes.to_date_type;
               vm.uiOptions.toDate = convertDateFormatFromServer(vm.leaveRequest.to_date);
-              vm.onDateChange(vm.uiOptions.toDate, 'to');
+              vm.onDateChange(vm.uiOptions.toDate, 'to')
+                .then( function () {
+                  //reolve only after both from and to day types are also set
+                  deferred.resolve({});
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: Do you need to pass any parameter at all if you really just want to `resolve` the promise?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 12]
```diff
@@ -829,7 +856,13 @@ define([
       function updateRequest() {
         vm.leaveRequest.update()
           .then(function () {
-            postSubmit('LeaveRequest::updatedByManager');
+            if (vm.role == 'manager') {
+              postSubmit('LeaveRequest::updatedByManager');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: What's the difference between "updatedByManager" and "edit"? Do we need two different events for the same edit functionality?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 13]
```diff
@@ -829,7 +856,13 @@ define([
...
+            }
+            else if (vm.role == 'owner') {
+              postSubmit('LeaveRequest::edit');
+            }
+
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: empty line

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js [hunk 14]
```diff
@@ -699,6 +699,91 @@
...
+          var directiveOptions = {
+            contactId: leaveRequest.contact_id, //owner's contact id
+            leaveRequest: leaveRequest
+          };
+
+          initTestController(directiveOptions);
+        });
+
+        describe('on initialization', function () {
+          var waiting_approval;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: do not use snake_case

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: Also please use more meaningful names, I didn't know what `waiting_approval` was until i read the initialization. You could simply this by just calling it `waitingApprovalStatus`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js [hunk 15]
```diff
@@ -699,6 +699,91 @@
...
+
+          beforeEach(function () {
+            waiting_approval = optionGroupMock.specificObject('hrleaveandabsences_leave_request_status', 'value', '3');
+          });
+
+          it('should set role to owner', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: do not use `should` in `it` titles. The title is always imperative
```js
it('sets the role to owner', function () {});
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js [hunk 16]
```diff
@@ -699,6 +699,91 @@
...
+            expect($ctrl.role).toEqual('owner');
+          });
+
+          it('should set all leaverequest values', function () {
+            expect($ctrl.leaveRequest.contact_id).toEqual('' + CRM.vars.leaveAndAbsences.contactId);
+            expect($ctrl.leaveRequest.type_id).toEqual(jasmine.any(String));
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: Why don't you use the actual values you have in the `leaveRequest` object? You are obviously testing if all the values are stored correctly in the controller, so testing for "any" string doesn't serve any purpose

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js [hunk 17]
```diff
@@ -699,6 +699,91 @@
...
+            expect($ctrl.leaveRequest.status_id).toEqual(waiting_approval.value);
+            expect($ctrl.leaveRequest.from_date).toEqual(jasmine.any(String));
+            expect($ctrl.leaveRequest.from_date_type).toEqual(jasmine.any(String));
+            expect($ctrl.leaveRequest.to_date).toEqual(jasmine.any(String));
+            expect($ctrl.leaveRequest.to_date_type).toEqual(jasmine.any(String));
+          });
+
+          it('will not allow user to submit', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: "does not allow"

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-report.html [hunk 18]
```diff
@@ -380,7 +380,8 @@
                               </a>
                               <ul class="dropdown-menu dropdown-menu-right" uib-dropdown-menu>
                                 <li ng-repeat="action in report.actionsFor(request)">
-                                  <a href ng-click="report.action(request, action)">{{action}}</a>
+                                  <a ng-if="action == 'cancel'" href ng-click="report.action(request, action)">{{action}}</a>
+                                  <a ng-if="action != 'cancel'" href leave-request-popup leave-request='request' contact-id="report.contactId">{{action}}</a>
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-report.html]: Do not use single quotes for any html attribute
```html
leave-request="request"
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 19]
```diff
@@ -285,7 +291,14 @@ define([
             initAbsenceType();
             initStatus();
             initContact();
-            initDates();
+
+            //the promise will set the day types so that initial value is set properly
+            return initDates();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: If `initDates` needs to be used to respect the order of execution while the other functions (`initAbsenceType`, `initStatus`, etc) don't, then it's better to separate it for readability sake
```js
.then(function () {
  return $q.all([loadDayTypes(), loadStatuses()]);
})
.then(function () {
  return initDates();
})
.then(function () {
  initAbsenceType();
  initStatus();
  initContact();
})

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js (hunk 2 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: that will be unnecessary call to server

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js (hunk 6 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: `initialLeaveRequest` is initialized after all init called

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js (hunk 12 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: the action maybe edit but the recipient are different, hence separate event, also manager could be owner for its leaverequest and that edit will be different from manager role edit

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js (hunk 15 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js]: should is also frequently used for many open source projects

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 20]
```diff
@@ -487,6 +481,20 @@ define([
       })['value'];
     }
 
+    /**
+    * Register events which will be called by other modules
+    *
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Empty line

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 21]
```diff
@@ -487,6 +481,20 @@ define([
...
+    */
+    function registerEvents() {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: Put the functions in alphabetical order

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 22]
```diff
@@ -288,16 +288,16 @@ define([
             return $q.all([loadDayTypes(), loadStatuses()]);
           })
           .then(function () {
+            //the promise will set the day types so that initial value is set properly
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]: This is not necessary, if the `initDates()` method is correctly documented

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js [hunk 23]
```diff
@@ -288,16 +288,16 @@ define([
...
             initStatus();
             initContact();
 
-            //the promise will set the day types so that initial value is set properly
-            return initDates();
-          })
-          .then( function () {
-            if(vm.mode == 'edit') {
-              initialLeaveRequest = _.cloneDeep(vm.leaveRequest.attributes());
+            if(vm.mode === 'edit') {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js]:
```js
if (vm.mode === 'edit') {
```

**Pull request merged**